subsys/usb/device_next/class/usbd_hid_api.c

```diff
@@ -0,0 +1,220 @@
+		.bRequest = 0,
+		.wValue = (type << 8) | id,
+		.wIndex = 0,
+		.wLength = len,
+	};
+	uint8_t *d = buf;
+	int l = len;
+
+	if (legacy_ops != NULL && legacy_ops->get_report != NULL) {
+		return legacy_ops->get_report(dev, &setup, &l, &d);
+	}
+
+	return -ENOTSUP;
+}
+
+int wrapper_set_report(const struct device *dev,
+		       const uint8_t type, const uint8_t id,
+		       const uint16_t len, const uint8_t *const buf)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+	struct usb_setup_packet setup = {
+		.bmRequestType = 0,
+		.bRequest = 0,
+		.wValue = (type << 8) | id,
+		.wIndex = 0,
+		.wLength = len,
+	};
+	uint8_t *d = (void *)buf;
+	int l = len;
+
+	if (legacy_ops != NULL && legacy_ops->set_report != NULL) {
+		return legacy_ops->get_report(dev, &setup, &l, &d);
+	}
+
+	return -ENOTSUP;
+}
+
+void wrapper_set_idle(const struct device *dev,
+		      const uint8_t id, const uint32_t duration)
+{
+	if (id != 0U) {
+		LOG_ERR("Set Idle for %p ID %u duration %u cannot be propagated",
+			dev, id, duration);
+	}
+}
+
+void wrapper_set_protocol(const struct device *dev, const uint8_t proto)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+
+	if (legacy_ops != NULL && legacy_ops->protocol_change != NULL) {
+		legacy_ops->protocol_change(dev, proto);
+	}
+}
+
+void wrapper_input_report_done(const struct device *dev)
+{
+	const struct hid_ops *legacy_ops = get_legacy_ops(dev);
+
+	if (legacy_ops == NULL) {
+		return;
+	}
+
+	if (legacy_ops->int_in_ready) {
+		LOG_ERR("Input report done for %p", dev);
+		legacy_ops->int_in_ready(dev);
+	}
+}
+
+void wrapper_output_report(const struct device *dev,
+			   const uint16_t len, const uint8_t *const buf)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(len);
+	ARG_UNUSED(buf);
+
+	__ASSERT(false, "Output report callback is not supported");
+}
+
+static struct hid_device_ops wrapper_ops = {
+	.get_report = wrapper_get_report,
+	.set_report = wrapper_set_report,
+	.set_idle = wrapper_set_idle,
+	.set_protocol = wrapper_set_protocol,
+	.input_report_done = wrapper_input_report_done,
+	.output_report = wrapper_output_report,
+};
+
+int hid_int_ep_write(const struct device *dev,
+		     const uint8_t *data, uint32_t data_len, uint32_t *bytes_ret)
+{
+	int ret;
+
+	ret = hid_device_submit_report(dev, data_len, data, false);
+	if (bytes_ret != NULL) {
+		*bytes_ret = ret == 0 ? data_len : 0;
+	}
+
+	return ret;
+}
+
+int hid_int_ep_read(const struct device *dev,
+		    uint8_t *data, uint32_t max_data_len, uint32_t *ret_bytes)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(data);
+	ARG_UNUSED(max_data_len);
+	ARG_UNUSED(ret_bytes);
+
+	LOG_ERR("Not supported");
+
+	return -ENOTSUP;
+}
+
+int usb_hid_set_proto_code(const struct device *dev, uint8_t proto_code)
+{
+	ARG_UNUSED(dev);
+	ARG_UNUSED(proto_code);
+
+	LOG_DBG("Protocol code is set using DT property boot-protocol-code");
+
+	return 0;
+}
+
+int usb_hid_init(const struct device *dev)
+{
+	LOG_DBG("It does nothing for dev %p", dev);
```

Comment:
Please cast the `%p` argument to `(void *)` (as required by C standard). Also in other places

Reply:
Required by what C standard?
I will not add explicit conversion of any pointer to a void pointer.

Reply:
Apparently all of the major ones. C89: "The argument shall be a pointer to void", C99 and C11: "The corresponding argument shall be a pointer to a pointer to void.", C11

Not respecting this leads to undefined behavior. Whether this is going to lead to actual real world problems is another story, but from the standard point of view it can (as any undefined behavior).

Reply:
The other way around is OK. The C standard guarantees conversion from `void*` to any type of pointer, but not in this direction as pointed out here.

Reply:
>The other way around is OK. The C standard guarantees conversion from void* to any type of pointer, but not in this direction as pointed out here.

That is different. C in a Nutshell, "Any object pointer type can be implicitly converted to a pointer to void, and vice versa.".
Not sure fprintf conversion specification is required for cbprintf.